src/vm/costs/mod.rs

```diff
@@ -93,37 +112,95 @@ impl CostTracker for () {
     fn reset_memory(&mut self) {}
 }
 
+#[derive(Debug, Deserialize, Serialize, Clone, PartialEq, Eq)]
+pub struct ClarityCostFunctionReference {
+    pub contract_id: QualifiedContractIdentifier,
+    pub function_name: String,
+}
+
+impl ClarityCostFunctionReference {
+    fn new(id: QualifiedContractIdentifier, name: String) -> ClarityCostFunctionReference {
+        ClarityCostFunctionReference { contract_id: id, function_name: name }
+    }
+}
+
+type ClarityCostContract = &'static str;
+
 #[derive(Debug, Clone, PartialEq)]
 pub struct LimitedCostTracker {
+    cost_function_references: HashMap<&'static ClarityCostFunction, ClarityCostFunctionReference>,
+    cost_contracts: HashMap<QualifiedContractIdentifier, ContractAST>,
     total: ExecutionCost,
     limit: ExecutionCost,
     memory: u64,
     memory_limit: u64,
+    free: bool,
 }
 
 #[derive(Debug, PartialEq, Eq)]
 pub enum CostErrors {
+    CostComputationFailed(String),
     CostOverflow,
     CostBalanceExceeded(ExecutionCost, ExecutionCost),
     MemoryBalanceExceeded(u64, u64),
 }
 
 impl LimitedCostTracker {
+    //TODO: require cost functions/contracts to be included
     pub fn new(limit: ExecutionCost) -> LimitedCostTracker {
-        LimitedCostTracker {
+        let mut cost_tracker = LimitedCostTracker {
+            cost_function_references: HashMap::new(),
+            cost_contracts: HashMap::new(),
             limit,
             memory_limit: CLARITY_MEMORY_LIMIT,
             total: ExecutionCost::zero(),
             memory: 0,
-        }
+            free: false,
+        };
+        //TODO: costs chould get passed in and this shouldn't be constructable without them
+        cost_tracker.load_boot_costs();
+        cost_tracker
     }
+    //TODO: require cost functions/contracts to be included
     pub fn new_max_limit() -> LimitedCostTracker {
+        let mut cost_tracker = LimitedCostTracker {
+            cost_function_references: HashMap::new(),
+            cost_contracts: HashMap::new(),
+            limit: ExecutionCost::max_value(),
+            total: ExecutionCost::zero(),
+            memory: 0,
+            memory_limit: CLARITY_MEMORY_LIMIT,
+            free: false,
+        };
+        cost_tracker.load_boot_costs();
+        cost_tracker
+    }
+    pub fn new_free() -> LimitedCostTracker {
```

Comment:
I've been thinking more about the implications of having a "free" cost tracker.  Since the only case where we would use a "free" cost-tracker is to parse and evaluate a cost function (risky, since it's a DDoS vector), might it make more sense to have a protocol-defined constant maximum cost for processing a cost function?

Reply:
We do free tracking in other places as well, like determining the PoX reward set, for example.

Reply:
Note also that the importance of performing free calculations isn't that there should be a maximum cost limit for processing a cost function, but instead the fact that otherwise we have a recursive problem: if we use a cost-tracking Clarity VM to compute costs, do we then use another cost-tracking Clarity VM to track cost of that VM? By making the actual computation of costs free, we escape that recursive problem. Of course, we want to be sure that these cost functions aren't going to be DDOS vectors, but we'd want to do that by limiting the space of those functions themselves to arithmetic (which is the plan).

Reply:
Right, understood about the need to bootstrap cost-tracking without recursion.  That's why I was wondering if we should consider having a tighter bound on "free" cost-tracking than the maximum possible value.

> but we'd want to do that by limiting the space of those functions themselves to arithmetic (which is the plan).

I asked about that (indirectly) below, regarding the use of `NullBackingStore`.  We'll want to limit ourselves to arithmetic for cost-tracking.

There's not a similar concern about PoX, right?  The PoX functions can't be removed or changed through a miner vote, so there's no need to worry about constraining the set of operations the PoX functions can do?

Reply:
> There's not a similar concern about PoX, right? The PoX functions can't be removed or changed through a miner vote, so there's no need to worry about constraining the set of operations the PoX functions can do?

Yep, that's right -- no need to constrain the set of operations for those functions, but we would want to execute them without executing the cost-tracking functions for efficiency.

Reply:
So, we're not worried about the free cost tracker being maximal?  I see that the rest of my concern in this issue is recorded in #2041.

Reply:
Nope-- not worried about the free cost tracker being maximal. Anywhere the free cost tracker is invoked should be getting locked down through other means.